server/controller/src/main/java/ai/starwhale/mlops/domain/dataset/DatasetService.java

```diff
@@ -371,4 +417,160 @@ public Map<String, String> signLinks(String project, String datasetName, Set<Str
+                rp = systemSettingService.getRunTimeProperties().getDatasetBuild().getResourcePool();
+            }
+            var pool = Objects.isNull(rp) ? null : systemSettingService.queryResourcePool(rp);
+            k8sJobTemplate.getContainersTemplates(job).forEach(templateContainer -> {
+                ContainerOverwriteSpec containerOverwriteSpec = new ContainerOverwriteSpec(templateContainer.getName());
+                containerOverwriteSpec.setEnvs(envVars);
+                containerOverwriteSpec.setImage(
+                        image.resolve(systemSettingService.getDockerSetting().getRegistryForPull()));
+                containerOverwriteSpec.setCmds(List.of("dataset_build"));
+                List<RuntimeResource> resources = Objects.isNull(pool) ? List.of()
+                        : pool.validateAndPatchResource(List.of(
+                            RuntimeResource.builder().type(RESOURCE_CPU).build(),
+                            RuntimeResource.builder().type(RESOURCE_MEMORY).build()));
+                log.info("using resource pool {}, patched resources {}", pool, resources);
+                containerOverwriteSpec.setResourceOverwriteSpec(new ResourceOverwriteSpec(resources));
+
+                ret.put(templateContainer.getName(), containerOverwriteSpec);
+            });
+            Map<String, String> nodeSelector = pool != null ? pool.getNodeSelector() : Map.of();
+            var toleration = pool != null ? pool.getTolerations() : null;
+            k8sJobTemplate.renderJob(job, String.format("%s-%d", entity.getDatasetName(), entity.getId()),
+                        "Never", 0, ret, nodeSelector, toleration, null);
+
+            log.debug("deploying dataset build job to k8s :{}", JSONUtil.toJsonStr(job));
+            try {
+                k8sClient.deployJob(job);
+            } catch (ApiException e) {
+                throw new SwProcessException(ErrorType.K8S, "deploy dataset build job failed", e);
+            }
+        } else {
+            throw new SwProcessException(ErrorType.DB, "create build record failed");
+        }
+    }
+
+    @Transactional
+    public boolean updateBuildStatus(Long id, BuildStatus status) {
+        var record = buildRecordMapper.selectById(id);
+        if (record == null) {
+            log.warn("build record:{} can't find when update status to {}.", id, status);
+            return false;
+        }
+        var res = buildRecordMapper.updateStatus(id, status) > 0;
+        if (res && status == BuildStatus.SUCCESS && record.getShared()) {
+            // update shared
+            var dataset = datasetMapper.findByName(record.getDatasetName(), record.getProjectId(), false);
+            if (dataset != null) {
+                var version = datasetVersionMapper.findByLatest(dataset.getId());
+                if (version != null) {
+                    datasetVersionMapper.updateShared(version.getId(), true);
+                }
+            }
```

Comment:
Make an issue to fix the race condition when the other user updates the dataset?

Reply:
ok, https://github.com/star-whale/starwhale/issues/2523